gobblin-example/src/main/java/gobblin/example/wikipedia/WikipediaExtractor.java

```diff
@@ -54,58 +64,87 @@
 
   private static final String JSON_MEMBER_QUERY = "query";
   private static final String JSON_MEMBER_PAGES = "pages";
   private static final String JSON_MEMBER_REVISIONS = "revisions";
   private static final String JSON_MEMBER_PAGEID = "pageid";
   private static final String JSON_MEMBER_TITLE = "title";
 
-  private static final Splitter SPLITTER = Splitter.on(",").omitEmptyStrings().trimResults();
 
   private static final Gson GSON = new Gson();
 
-  private final WorkUnitState workUnitState;
-  private final WorkUnit workUnit;
   private final WikiResponseReader reader;
-  private final int revisionsCnt;
   private final String rootUrl;
   private final String schema;
-  private final Queue<String> requestedTitles;
-  private final int numRequestedTitles;
-  private Queue<JsonElement> recordsOfCurrentTitle;
+  private final String requestedTitle;
+  private final int batchSize;
+  private final long lastRevisionId;
+  private Queue<JsonElement> currentBatch;
+  private final ImmutableMap<String, String> baseQuery;
+  private final WorkUnitState workUnitState;
+  private final int maxRevisionsPulled;
 
   private class WikiResponseReader implements Iterator<JsonElement> {
 
+    private long latestPulledRevision;
+    private long revisionsPulled = 0;
+
+    public WikiResponseReader(long latestPulledRevision) {
+      this.latestPulledRevision = latestPulledRevision;
+    }
+
     @Override
     public boolean hasNext() {
-      if (!WikipediaExtractor.this.recordsOfCurrentTitle.isEmpty()) {
-        return true;
-      } else if (WikipediaExtractor.this.requestedTitles.isEmpty()) {
+
+      if (WikipediaExtractor.this.maxRevisionsPulled > -1
+          && this.revisionsPulled >= WikipediaExtractor.this.maxRevisionsPulled) {
+        WikipediaExtractor.this.workUnitState.setActualHighWatermark(new LongWatermark(this.latestPulledRevision));
+        LOG.info(String.format("Pulled max number of records %d, final revision pulled %d.", this.revisionsPulled,
+            this.latestPulledRevision));
         return false;
+      }
+
+      if (!WikipediaExtractor.this.currentBatch.isEmpty()) {
+        return true;
       } else {
 
         /*
          * Retrieve revisions for the next title. Repeat until we find a title that has at least one revision,
          * otherwise return false
          */
-        while (!WikipediaExtractor.this.requestedTitles.isEmpty()) {
-          String currentTitle = WikipediaExtractor.this.requestedTitles.poll();
-          try {
-            WikipediaExtractor.this.recordsOfCurrentTitle = retrievePageRevisions(currentTitle);
-          } catch (IOException e) {
-            LOG.error("IOException while retrieving revisions for title '" + currentTitle + "'");
-          }
-          if (!WikipediaExtractor.this.recordsOfCurrentTitle.isEmpty()) {
-            return true;
-          }
+        if (this.latestPulledRevision >= WikipediaExtractor.this.lastRevisionId) {
+          return false;
```

Comment:
Because `actualHighWatermark` is set in the constructor. It needs to be updated in lines 112-114 only if we don't actually reach that watermark.